PR: Display interpreter status if Spyder not launched from conda envrionment (Mac app)

mrclary opened the pull request:
* display interpreter status in statusbar, even if it isn't a conda interpreter or spyder isn't started from conda environment.

<!--- Make sure to read the Contributing Guidelines:                   --->
<!--- https://github.com/spyder-ide/spyder/blob/master/CONTRIBUTING.md --->
<!--- and follow PEP 8, PEP 257 and Spyder's code style:               --->
<!--- https://github.com/spyder-ide/spyder/wiki/Dev:-Coding-Style      --->

## Description of Changes
The conda status bar no longer tests if Spyder's python interpreter is from a conda environment.

* [x] Wrote at least one-line docstrings (for any new functions)
* [x] Added unit test(s) covering the changes (if testable)
<!--- Remember that an image/animation is worth a thousand words! --->
* [x] Included a screenshot or animation (if affecting the UI, see [Licecap](https://www.cockos.com/licecap/))


<!--- Explain what you've done and why --->

Currently, when updating the conda status in the status bar, Spyder's python environment is tested for being a conda environment rather than the iPython console's interpreter environment.
When Spyder is launched from a non-conda environment (such as in a standalone Mac application), the iPython console interpreter status is not displayed.
This PR fixes that.


### Issue(s) Resolved

<!--- List the issue(s) below, in the form "Fixes #1234"; one per line --->

Fixes spyder-ide/mac-application#6

### Affirmation

By submitting this Pull Request or typing my (user)name below,
I affirm the [Developer Certificate of Origin](https://developercertificate.org)
with respect to all commits and content included in this PR,
and understand I am releasing the same under Spyder's MIT (Expat) license.

<!--- TYPE YOUR USER/NAME AFTER THE FOLLOWING: --->
I certify the above statement is true and correct:
@mrclary
<!--- Thanks for your help making Spyder better for everyone! --->

pep8speaks commented:
Hello @mrclary! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`spyder/widgets/status.py`](https://github.com/spyder-ide/spyder/blob/84ac91b5de47b67c6e1266cfdb19e8eddc07e72f/spyder/widgets/status.py):

> [Line 283:1](https://github.com/spyder-ide/spyder/blob/84ac91b5de47b67c6e1266cfdb19e8eddc07e72f/spyder/widgets/status.py#L283): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

ccordoba12 commented:
@goanpeca, please review this one.

goanpeca commented:
@mrclary I think the first thing we need to fix is the naming, if this is no longer a CondaStatusWidget then having that name is misleading.

Could you adapt the code to handle this renaming?

goanpeca commented on spyder/widgets/status.py:
```diff
@@ -254,21 +254,23 @@ def _process_conda_env_info(self):
-        if envs_folder in self._interpreter:
-            if os.name == 'nt':
-                env = os.path.dirname(self._interpreter)
+        if 'conda' in self._interpreter:
```
I don't know if this is robust enough

mrclary replied:
Do you have a recommendation? I'm happy to improve it, if I can. Is it any less robust than `utils.is_anaconda` or `envs_folder in self._interpreter`?.

goanpeca replied:
> Is it any less robust than utils.is_anaconda

Yes it is way less robust that `is_anaconda`, so please fix this

mrclary replied:
> Not sure why you stopped using a function with a clear purpose.

Because it's the wrong function to use here. `utils.is_anaconda` simply checks if `sys.prefix` is a conda environment; this is only Spyder's environment and may or may not be the same environment as `self._interpreter`. But I think I understand your concerns. I've changed it to use `spyder.util.is_conda_env` instead.

goanpeca replied:
Yes, I meant that it needed a small refactor to check if it was a conda env.

The same proces that checks for the existence of the file in sys.prefix, applied to the conda env.

mrclary commented:
@goanpeca , thanks for reviewing this PR.
> Also the renaming of Conda... to PythonInterpreterStatusWidget might be better if we want to generalize.
> 
> Eventually I guess other interpreters might be used (for other languages when they come) but that will be work for another PR.

I can certainly rename relevant classes and functions. Related is feature request #12200.

mrclary commented:
@goanpeca , I'm having some trouble setting up the mock object correctly in the test that's failing. I've traded `is_anaconda` for `is_conda_env` in the status bar widget and have tried to modify the mock object accordingly in the test. It returns the correct value in the test namespace but not the correct value in the `_process_interpreter_env_info` namespace. What am I doing wrong?

goanpeca commented:
@mrclary I will take a look at what might be the problem.

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -872,11 +872,10 @@ def create_edit_action(text, tr_text, icon):
+        from spyder.widgets.status import InterpreterStatus
+        self.conda_status = InterpreterStatus(self, status,
+                                              icon=ima.icon('environment'))
+        self.conda_status.update_interpreter(self.get_main_interpreter())
```
We should also change the name of this variable to 

`self.interpreter_status`

mrclary replied:
Will do.

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -3120,7 +3119,7 @@ def apply_statusbar_settings(self):
 
             # Update conda status widget
-            if is_anaconda() and self.conda_status:
+            if self.conda_status:
```
same here

mrclary replied:
Will do.

mrclary commented:
> @mrclary I will take a look at what might be the problem.

@goanpeca , have you had a chance to look at this yet? I know you're real busy...

goanpeca commented:
Sorry @mrclary I don't think I will be able to look at this until the weekend

mrclary commented:
@goanpeca , I could use your input on the mock issue here. I noticed that you marked this for the April sprint milestone. I'm not in any hurry, so I don't want to press anything when you likely have higher priorities :-)

goanpeca commented on spyder/app/mainwindow.py:
```diff
@@ -3139,9 +3138,9 @@ def apply_statusbar_settings(self):
                     widget.set_interval(CONF.get('main', '%s/timeout' % name))
 
             # Update conda status widget
```
We need to update this comment

mrclary replied:
can do

goanpeca commented on spyder/config/base.py:
```diff
@@ -536,7 +536,7 @@ def translate_gettext(x):
-def running_in_mac_app():
+def running_in_mac_app(pyexec=None):
     """
     Check if Spyder is running inside an app on macOS.
```
Let's add some additional info on the docstring on the pyexec paramater

mrclary replied:
can do

goanpeca commented on spyder/widgets/tests/test_status.py:
```diff
@@ -62,53 +65,70 @@ def test_status_bar_widget_signal(status_bar, qtbot):
-def test_status_bar_conda_status(status_bar, qtbot, mocker):
-    mocker.patch.object(utils, 'is_anaconda', return_value=True)
+def test_status_bar_conda_interpreter_status(status_bar, qtbot, mocker):
+    # We patch where the method is used not where it is imported from
```
@mrclary this is why the mocking was not working, although not intuitive at first, we need to patch the method where we use it, not where it comes from.

mrclary replied:
Aahh, yes, of course. I'll fix that...

goanpeca replied:
Already did it for you! :-)

goanpeca commented:
@mrclary thanks again, and sorry for taking so long to review and help with the failing test. I rebased with master and added a commit to fix the test and some style issues. Sorry in the rebase process I "co authored" your previous commits 🙃

mrclary commented:
> @mrclary thanks again, and sorry for taking so long to review and help with the failing test. I rebased with master and added a commit to fix the test and some style issues. Sorry in the rebase process I "co authored" your previous commits 🙃

No problem! Thanks for your help. I'll work around your rebase, if needed.

goanpeca commented on spyder/widgets/tests/test_status.py:
```diff
@@ -62,53 +66,71 @@ def test_status_bar_widget_signal(status_bar, qtbot):
+
+@pytest.mark.skipif(sys.platform != 'darwin', reason="Only valid on Mac")
+def test_status_bar_internal_interpreter_status(status_bar, qtbot, mocker):
+    mocker.patch.object(conda, 'is_conda_env', return_value=False)
```
@mrclary so something like this:

```
    mocker.patch.object(spyder.widgets.status, 'is_conda_env',
                        return_value=False)
```

goanpeca commented:
Wohoo test pass :-) !

mrclary commented:
Thanks for all the help on this @goanpeca.

ccordoba12 commented on spyder/widgets/tests/test_status.py:
```diff
@@ -62,53 +66,75 @@ def test_status_bar_widget_signal(status_bar, qtbot):
-    mocker.patch.object(utils, 'is_anaconda', return_value=True)
+def test_status_bar_conda_interpreter_status(status_bar, qtbot, mocker):
+    # We patch where the method is used not where it is imported from
+    mocker.patch.object(spyder.widgets.status, 'is_conda_env',
```
`is_conda_env` is a function of `spyder.utils.conda`. If possible, I think it's better to import it from that module directly instead of patching it indirectly through the import present in `spyder.widgets.status`.

mrclary replied:
Hmm..., @goanpeca advised the current implementation because we couldn't get mocker to patch it when we imported from `spyder.utils.conda`, hence the comment `# We patch where the method is used not where it is imported from`.  
See resolved conversation above: https://github.com/spyder-ide/spyder/pull/12232#discussion_r421917790

If we missed something there, please let me know.

ccordoba12 replied:
Sorry, I missed that comment. It's a minor, nitpicky detail. The most important thing is to test this appropriately, which you did.

ccordoba12 commented on spyder/widgets/tests/test_status.py:
```diff
@@ -62,53 +66,75 @@ def test_status_bar_widget_signal(status_bar, qtbot):
-def test_status_bar_no_conda_status(status_bar, qtbot, mocker):
-    mocker.patch.object(utils, 'is_anaconda', return_value=False)
+def test_status_bar_other_interpreter_status(status_bar, qtbot, mocker):
+    mocker.patch.object(spyder.widgets.status, 'is_conda_env',
```
The same previous suggestion applies here.

ccordoba12 commented on spyder/widgets/tests/test_status.py:
```diff
@@ -62,53 +66,75 @@ def test_status_bar_widget_signal(status_bar, qtbot):
+
+@pytest.mark.skipif(sys.platform != 'darwin', reason="Only valid on Mac")
+def test_status_bar_internal_interpreter_status(status_bar, qtbot, mocker):
+    mocker.patch.object(spyder.widgets.status, 'is_conda_env',
```
And here.

ccordoba12 commented on spyder/widgets/tests/test_status.py:
```diff
@@ -62,53 +66,75 @@ def test_status_bar_widget_signal(status_bar, qtbot):
+def test_status_bar_internal_interpreter_status(status_bar, qtbot, mocker):
+    mocker.patch.object(spyder.widgets.status, 'is_conda_env',
+                        return_value=False)
+    mocker.patch.object(spyder.widgets.status, 'running_in_mac_app',
```
It should also be possible to patch `running_in_mac_app` by importing it from the module it is defined.

ccordoba12 reviewed (commented):
Thanks @mrclary for this! It looks pretty good!

I only left some small comments in tests touched by your work.

ccordoba12 commented:
Also, this has a conflict with 4.x now.

mrclary commented:
> Also, this has a conflict with 4.x now.

No problem, I'll take care of it.

I should mention that the following PRs should be addressed in order:

1. #12011 
1. #12232 
1. #12235 
1. #12236

ccordoba12 commented:
> I should mention that the following PRs should be addressed in order

Ok, thanks for letting me know about it. Let's work on #12011 then.

ccordoba12 reviewed (commented):
@mrclary, I think this one is ready. Please rebase to remove the commits of PR #12011 from here, so I can merge it.

mrclary commented:
@ccordoba12, it appears that "Install dependencies" failed. Is there a way to abort the tests and restart them?

ccordoba12 commented:
I restarted the failing slots.

ccordoba12 reviewed (approved):
Thanks @mrclary!
